### ISSUE TITLE
cuda graph capturing failing

### ISSUE THREAD

**Issue by stephen-youn** [HazyResearch/flash-attention#312]
Hi,
I found that the unpad_input function make the cuda capture failing if we want to add key_attention_mask.

https://github.com/HazyResearch/flash-attention/blob/72ad03eaa661f6bf3a14c855316c27fbab4f8f4c/flash_attn/bert_padding.py#L108C1-L108C80

called from here.
[
](https://github.com/HazyResearch/flash-attention/blob/72ad03eaa661f6bf3a14c855316c27fbab4f8f4c/flash_attn/flash_attention.py#L26C28-L26C49)

is there a quick way to make this cuda-capturable ?

thanks

**Comment by tridao** [comment]
I don't think so, since the size of the output isn't known before hand (i.e. it depends on the key_padding_mask).
For MLPerf, Nvidia uses CUDA graph for some of the submissions and they don't remove the padding tokens. The tradeoff there is you spend more compute time processing the padding tokens, but reduces the overhead with CUDA graph. This can make sense for very small batch sizes (e.g. I think Nvidia was running with batch size 3-4 on thousands of GPUs) where the overhead is larger than the wasted compute time processing the padding tokens.
